### PR TITLE
build: modularize the gulp file to be easier to maintain

### PR DESCRIPTION
**Pull request by petebacondarwin**
This is a precursor to bringing in some of the gulp tasks from angular.io,
which could lead to the gulpfile becoming unwieldy.

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[ ] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

**Comment by petebacondarwin** [comment]
Here is some background to this approach: http://macr.ae/article/splitting-gulpfile-multiple-files.html

**Review by IgorMinar — CHANGES_REQUESTED** [review]
- commit message: "build: ...."
- as discussed let's move requires into closures so that they execute lazily.

File: gulpfile.js [hunk 1]
```diff
@@ -16,271 +16,29 @@ require('./tools/check-environment')({
...
-
-gulp.task('docs-test', ['doc-gen-test', 'docs-app-test']);
-gulp.task('doc-gen-test', () => {
-  const execSync = require('child_process').execSync;
-  execSync(
-      'node dist/tools/cjs-jasmine/index-tools ../../tools/docs/**/*.spec.js',
-      {stdio: ['inherit', 'inherit', 'inherit']});
-});
+gulp.task('doc-gen', loadTask('docs', 'generate'));
+gulp.task('doc-gen-test', () => loadTask('docs', 'test'));
```
Comment (gkalpak) [inline gulpfile.js]: Shouldn't this be `loadTask(...)`, instead of `() => loadTask(...)`?

**Review by gkalpak — COMMENTED** [review]
If we wanted to take "lazy-requiring" to an extreme, couldn't we avoid rewuiring the tasks until they are actually run? I.e. have `loadTask` return a _function_ that loads and runs the task?

File: gulpfile.js (hunk 1 above)
Comment (petebacondarwin) [inline gulpfile.js]: yes!

**Review by IgorMinar — APPROVED** [review]

**Pull request merged**